src/main/java/org/jboss/netty/channel/socket/nio/AbstractNioChannel.java

```diff
@@ -256,10 +257,14 @@ public boolean offer(MessageEvent e) {
             if (newWriteBufferSize >= highWaterMark) {
                 if (newWriteBufferSize - messageSize < highWaterMark) {
                     highWaterMarkCounter.incrementAndGet();
-                    if (!notifying.get() && setUnwritable()) {
-                        notifying.set(Boolean.TRUE);
-                        fireChannelInterestChanged(AbstractNioChannel.this);
-                        notifying.set(Boolean.FALSE);
+                    if (setUnwritable()) {
+                        if (!isIoThread(AbstractNioChannel.this)) {
+                            fireChannelInterestChangedLater(AbstractNioChannel.this);
+                        } else if (!notifying.get()) {
```

Comment:
bummer Ijust merged as -i thought from your comment before that it is ok as it is now :/ sorry